azimuth_identity/operator.py

```diff
@@ -187,11 +195,129 @@ async def delete_realm(instance: api.Realm, **kwargs):
+@model_handler(api.OIDCClient, kopf.on.update, field = "spec", param = "UPDATE")
+@model_handler(api.OIDCClient, kopf.on.resume, param = "RESUME")
+async def reconcile_oidc_client(instance: api.OIDCClient, param, **kwargs):
+    """
+    Handles the reconciliation of an OIDC client.
+    """
+    # Acknowledge the client at the earliest opportunity
+    if instance.status.phase == api.OIDCClientPhase.UNKNOWN:
+        instance.status.phase = api.OIDCClientPhase.PENDING
+        LOGGER.info("Updating OIDC client status to PENDING - %s", format_instance(instance))
+        await save_instance_status(instance)
+    # If the spec has changed, put the client into the updating phase
+    if param == "UPDATE" and instance.status.phase != api.OIDCClientPhase.UPDATING:
+        instance.status.phase = api.OIDCClientPhase.UPDATING
+        LOGGER.info("Updating OIDC client status to UPDATING - %s", format_instance(instance))
+        await save_instance_status(instance)
+    # Get the realm for the client and wait for it to become ready
+    LOGGER.info("Fetching realm for OIDC client - %s", format_instance(instance))
+    realm = await wait_for_realm(instance.spec.realm_name, instance.metadata.namespace)
+    realm_name = keycloak.realm.realm_name(realm)
+    client_id, client_secret = await keycloak.oidc_client.ensure_oidc_client(realm_name, instance)
+    # If the client has a secret, write the client ID and secret to a Kubernetes secret
+    # If the client does NOT have a secret, delete the Kubernetes secret
+    # The OIDC client owns the secret so it gets deleted when the client does
+    secrets = await ekclient.api("v1").resource("secrets")
+    secret_name = (
+        instance.spec.credentials_secret_name or
+        f"{instance.metadata.name}-oidc-credentials"
+    )
+    if client_secret:
+        await secrets.server_side_apply(
+            secret_name,
+            {
+                "metadata": {
+                    "labels": {
+                        "app.kubernetes.io/managed-by": "azimuth-identity-operator",
+                    },
+                    "ownerReferences": [
+                        {
+                            "apiVersion": instance.api_version,
+                            "kind": instance.kind,
+                            "name": instance.metadata.name,
+                            "uid": instance.metadata.uid,
+                            "controller": True,
+                            "blockOwnerDeletion": True,
+                        },
+                    ],
+                },
+                "stringData": {
+                    "issuer-url": realm.status.oidc_issuer_url,
+                    "client-id": client_id,
+                    "client-secret": client_secret,
+                },
+            },
+            namespace = instance.metadata.namespace,
+            force = True
+        )
+    else:
+        await secrets.delete(secret_name, namespace = instance.metadata.namespace)
```

Comment:
We should maybe add a log message a head of this delete call, just for context.

What is the case where we would want to remove the secret?

Reply:
OK, this case is where we move from confidential to public, and the secret is removed.